vernac/declare.ml

```diff
@@ -172,6 +223,88 @@ type constant_entry =
   | PrimitiveEntry of primitive_entry
   | SymbolEntry of symbol_entry
 
+module ProofEntry = struct
+
+  let map_entry_body ~f = function
+    | Default { body = body; opaque } -> Default { body = f body; opaque }
+    | DeferredOpaque { body; feedback_id } ->
+      let body = Future.chain body (fun ((b,c),eff) -> let b, eff = f (b,eff) in ((b,c),eff)) in
+      DeferredOpaque { body; feedback_id }
+
+  let map_proof_entry ~f entry =
+    { entry with proof_entry_body = map_entry_body ~f entry.proof_entry_body }
+
+  let map_entry_type ~f entry =
+    { entry with proof_entry_type = f entry.proof_entry_type }
+
+  let get_opacity entry =
+    match entry.proof_entry_body with
+    | Default { body; opaque = Transparent } -> false
+    | Default { body; opaque = Opaque _ } -> true
+    | DeferredOpaque _ -> true
+
+  let force_entry_body entry =
+    match entry.proof_entry_body with
+    | Default { body; opaque } -> body, opaque
+    | DeferredOpaque { body; feedback_id = Some _ } -> CErrors.anomaly (str "Forcing a proof with feedback")
```

Comment:
I think you should enforce this statically, since several callers below are using `Future.from_val` for no good reason.

Reply:
It remains 3 `Future.from_val` in the current PR:
- one is to use `cast_opaque_proof_entry` and we could have a third case to it to avoid the from_val,
- one is to declare an opaque `Let`: I don't understand well the code but it can probably be avoided,
- the third one is to declare deferred univs for a non-deferred constant, and this could be avoided too.

So, all three could be avoided (assuming you are ok to have a third variant of `cast_opaque_proof_entry`).

Regarding `force_entry_body`, this is used for obligations and for `Derive`. I have the feeling that we rely here on stm invariants, that is that the stm will not defer proofs of obligations and proofs of `Derive`. I don't know how to ensure that (except by duplicating or annotating the type `proof_object` so that `close_proof` and `close_future_proof` have different return types). In any cases, it would modify the API of `declare.ml` and I'm not very comfortable at doing it now, at least until the whole coq/ceps#89 is in place.

Reply:
> third variant of cast_opaque_proof_entry

GADTs can do wonder to prevent code duplication.

>  I'm not very comfortable at doing it now

No problem, it's better to do it incrementally.

Reply:
If one wants to avoid futures as much as possible, one would also need a version of `Opaque.declare_defined_opaque` that takes a non deferred body. Do you want that?

Reply:
...or a dependently-typed version of `Opaque.declare_defined_opaque`.